File: Rule Engine/Services/Axis.Service.Scheduling/GroupScheduling/GroupSchedulingService.cs
using System.Collections.Generic;
using System.Collections.Specialized;
using Axis.Configuration;
using Axis.Model.Common;
using Axis.Model.Scheduling;
using Axis.Security;
using System.Globalization;

namespace Axis.Service.Scheduling.GroupScheduling
{
    public class GroupSchedulingService : IGroupSchedulingService
    {
        #region Class Variables

        private readonly CommunicationManager _communicationManager;
        private const string BaseRoute = "groupscheduling/";

        #endregion

        #region Constructors

        public GroupSchedulingService()
        {
            _communicationManager = new CommunicationManager(ApplicationSettings.Token, AuthContext.Auth.Token.Token);
        }

        #endregion

        #region Public Methods

        public Response<GroupSchedulingModel> GetGroupByID(long groupID)
        {
            string apiUrl = BaseRoute + "GetGroupByID";
            var param = new NameValueCollection { { "groupID", groupID.ToString() } };

            return _communicationManager.Get<Response<GroupSchedulingModel>>(param, apiUrl);
        }

        public Response<GroupSchedulingResourceModel> GetGroupSchedulingResource(long groupID)
        {
            string apiUrl = BaseRoute + "GetGroupSchedulingResource";
            var param = new NameValueCollection { { "groupID", groupID.ToString() } };

            return _communicationManager.Get<Response<GroupSchedulingResourceModel>>(param, apiUrl);
        }

        public Response<AppointmentModel> GetAppointmentByGroupID(long groupID)
        {
            string apiUrl = BaseRoute + "GetAppointmentByGroupID";
            var param = new NameValueCollection { { "groupID", groupID.ToString() } };

            return _communicationManager.Get<Response<AppointmentModel>>(param, apiUrl);
        }

        public Response<AppointmentContactModel> GetAllContactResourceNamesByAppointment(long appointmentID)
        {
            string apiUrl = BaseRoute + "GetAllContactResourceNamesByAppointment";
            var param = new NameValueCollection { { "appointmentID", appointmentID.ToString() } };

            return _communicationManager.Get<Response<AppointmentContactModel>>(param, apiUrl);
        }

        public Response<GroupSchedulingModel> AddGroupData(GroupSchedulingModel group)
        {
            string apiUrl = BaseRoute + "AddGroupData";
            return _communicationManager.Post<GroupSchedulingModel, Response<GroupSchedulingModel>>(group, apiUrl);
        }

        public Response<GroupSchedulingResourceModel> AddResources(List<GroupSchedulingResourceModel> resources)
        {
            string apiUrl = BaseRoute + "AddResources";
            return _communicationManager.Post<List<GroupSchedulingResourceModel>, Response<GroupSchedulingResourceModel>>(resources, apiUrl);
        }

        public Response<GroupSchedulingModel> UpdateGroupData(GroupSchedulingModel group)
        {
            string apiUrl = BaseRoute + "UpdateGroupData";
            return _communicationManager.Post<GroupSchedulingModel, Response<GroupSchedulingModel>>(group, apiUrl);
        }

        public Response<GroupSchedulingResourceModel> UpdateResources(List<GroupSchedulingResourceModel> resources)
        {
            string apiUrl = BaseRoute + "UpdateResources";
            return _communicationManager.Post<List<GroupSchedulingResourceModel>, Response<GroupSchedulingResourceModel>>(resources, apiUrl);
        }

        public Response<GroupSchedulingResourceModel> DeleteGroupSchedulingResource(long id, System.DateTime modifiedOn)
        {
            const string apiUrl = BaseRoute + "DeleteGroupSchedulingResource";
            var requestXMLValueNvc = new NameValueCollection
            {
                {"id", id.ToString(CultureInfo.InvariantCulture) },
                {"modifiedOn", modifiedOn.ToString(CultureInfo.InvariantCulture)}
            };
            return _communicationManager.Delete<Response<GroupSchedulingResourceModel>>(requestXMLValueNvc, apiUrl);
        }

        #endregion
    }
}

File: Presentation Engine/Axis.Plugins/Axis.Plugins.Scheduling/Scripts/app/services/groupSchedulingService.js
angular.module('xenatixApp')
    .factory('groupSchedulingService', ["$http", "$q", 'settings', function ($http, $q, settings) {
        var CONFIG = {
            controllerAction: "/data/plugins/scheduling/groupscheduling/"
        };

        function getGroupByID(groupID) {
            var dfd = $q.defer();
            $http.get(settings.webApiBaseUrl + CONFIG.controllerAction + 'getGroupByID', { params: { GroupID: groupID } })
            .success(function (data, status, header, config) {
                dfd.resolve(data);
            })
            .error(function (data, status, header, config) {
                dfd.reject(status);
            });

            return dfd.promise;
        };

        function getGroupSchedulingResource(groupID) {
            var dfd = $q.defer();
            $http.get(settings.webApiBaseUrl + CONFIG.controllerAction + 'getGroupSchedulingResource', { params: { GroupID: groupID } })
            .success(function (data, status, header, config) {
                dfd.resolve(data);
            })
            .error(function (data, status, header, config) {
                dfd.reject(status);
            });

            return dfd.promise;
        };

        function getAppointmentByGroupID(groupID) {
            var dfd = $q.defer();
            $http.get(settings.webApiBaseUrl + CONFIG.controllerAction + 'getAppointmentByGroupID', { params: { GroupID: groupID } })
            .success(function (data, status, header, config) {
                dfd.resolve(data);
            })
            .error(function (data, status, header, config) {
                dfd.reject(status);
            });

            return dfd.promise;
        };

        function getAllContactResourceNamesByAppointment(appointmentID) {
            var dfd = $q.defer();
            $http.get(settings.webApiBaseUrl + CONFIG.controllerAction + 'GetAllContactResourceNamesByAppointment', { params: { appointmentID: appointmentID } })
            .success(function (data, status, header, config) {
                dfd.resolve(data);
            })
            .error(function (data, status, header, config) {
                dfd.reject(status);
            });

            return dfd.promise;
        };

        function addGroupData(group) {
            var dfd = $q.defer();
            $http.post(settings.webApiBaseUrl + CONFIG.controllerAction + 'AddGroupData', group)
            .success(function (data, status, header, config) {
                dfd.resolve(data);
            })
            .error(function (data, status, header, config) {
                dfd.reject(status);
            });

            return dfd.promise;
        };

        function addResources(resources) {
            var dfd = $q.defer();
            $http.post(settings.webApiBaseUrl + CONFIG.controllerAction + 'AddResources', resources)
            .success(function (data, status, header, config) {
                dfd.resolve(data);
            })
            .error(function (data, status, header, config) {
                dfd.reject(status);
            });

            return dfd.promise;
        };

        function updateGroupData(group) {
            var dfd = $q.defer();
            $http.post(settings.webApiBaseUrl + CONFIG.controllerAction + 'UpdateGroupData', group)
            .success(function (data, status, header, config) {
                dfd.resolve(data);
            })
            .error(function (data, status, header, config) {
                dfd.reject(status);
            });

            return dfd.promise;
        };

        function updateResources(resources) {
            var dfd = $q.defer();
            $http.post(settings.webApiBaseUrl + CONFIG.controllerAction + 'UpdateResources', resources)
            .success(function (data, status, header, config) {
                dfd.resolve(data);
            })
            .error(function (data, status, header, config) {
                dfd.reject(status);
            });

            return dfd.promise;
        };

        function deleteGroupSchedulingResource(groupId, groupResourceId) {
            var dfd = $q.defer();
            $http.delete(settings.webApiBaseUrl + CONFIG.controllerAction + 'deleteGroupSchedulingResource', { params: { id: groupResourceId, modifiedOn: moment.utc().format() } })
                .success(function (data, status, header, config) {
                    dfd.resolve(data);
                })
                .error(function (data, status, header, config) {
                    dfd.reject(status);
                });

            return dfd.promise;
        };

        return {
            getGroupByID: getGroupByID,
            getGroupSchedulingResource: getGroupSchedulingResource,
            getAppointmentByGroupID: getAppointmentByGroupID,
            getAllContactResourceNamesByAppointment: getAllContactResourceNamesByAppointment,
            addGroupData: addGroupData,
            addResources: addResources,
            updateGroupData: updateGroupData,
            updateResources: updateResources,
            deleteGroupSchedulingResource: deleteGroupSchedulingResource
        };
    }]);